### ISSUE TITLE
[API Proposal]: ObjectAssertions Class Should Allow Only One Generic Type

### ISSUE THREAD

**Issue by JordanW9232** [fluentassertions/fluentassertions#2755]
### Background and motivation

Consider the following extension method:
```
public static class ObjectAssertionsExtensions
{
  public static AndConstraint<ObjectAssertions<T>> Should<T>(this T instance, Action<T> should)
  {
    should(instance);

    var assertions = new ObjectAssertions<T>(instance);
    return new AndConstraint<ObjectAssertions<T>>(assertions);
  }
}
```

Unfortunately, this method as written will not work since ObjectAssertions requires 2 generic types, but the second generic type is always just a reference to itself, so the class itself should be able to do this out of the box.

### API Proposal

```C#
public class ObjectAssertions<T> : ObjectAssertions<T, ObjectAssertions<T>>
{
  public ObjectAssertions(T value)
    : base(value)
  { }
}
```

**Note:**
For now, I've implemented this myself in my code and it works with my extension method, but I think it should be a part of the library rather than implementing it myself.

### API Usage

```C#
var result = new
{
  Id = 1,
  Name = "Test"
};

result.Should(x =>
{
  x.Id.Should().Be(1);
  x.Name.Should().Be("Test");
});
```

### Alternative Designs

Maybe consider adding the extension method I have in the library as well? I think it's useful enough to be used by many people.

### Risks

This shouldn't have any risks involved.

Even if the extension method I mentioned was added and also is named `Should`, it is a different signature so shouldn't cause any issues.

### Are you willing to help with a proof-of-concept (as PR in that or a separate repo) first and as pull-request later on?

Yes, please assign this issue to me.

**Comment by ITaluone** [comment]
I think this sounds really much like the new `ReferenceTypeAssertions.Satisfy<T>()` which has been implemented in https://github.com/fluentassertions/fluentassertions/pull/2597 and will be part of v7.

**Comment by JordanW9232** [comment]
> I think this sounds really much like the new `ReferenceTypeAssertions.Satisfy<T>(Action<T>)` which has been implemented in #2597 and will be part of v7.

Yeah I just looked and it totally does. That's enough to suit what I wanted to do. Thanks!

Although my only feedback on that is I think it's odd I have to specify the type like `Satisfy<MyType>` when I actually already know my type. It's only because you lose the specific type when calling `Should()` that you need to specify that generic

**Comment by ITaluone** [comment]
Huh.. I think this is a technical limitation. If I remember correctly, this has something to with overload resolution of the generic `Should`.

**Comment by dennisdoomen** [comment]
> Huh.. I think this is a technical limitation. If I remember correctly, this has something to with overload resolution of the generic `Should`.

That is correct. As soon as you define a generic version of `Should`, it'll supersede all the other overloads. The only way to make that work is to make ALL other assertion methods extension methods on the object that `Should<T>` returns. 

That's not what we plan to do.
